addons/netfox/rollback/rollback-synchronizer.gd

```diff
@@ -335,50 +323,40 @@ func _submit_full_state(received_state: Dictionary, received_tick: int):
 		
 		sanitized[property_name] = value
 	
-	# Duplicates the previous state(s), including the current one
-	# Also fixes packet loss (e.g. a state missing)
-	if (NetworkRollback.enable_state_diffs && _latest_state_tick != -1):
-		for picked_tick in range(_latest_state_tick, received_tick - 1):
-			_states[picked_tick + 1] = _states[picked_tick].duplicate(true)
-	
-	if sanitized.size() > 0:
-		_states[received_tick] = PropertySnapshot.merge(_states.get(received_tick, {}), sanitized)
-		_latest_state_tick = received_tick
+	if sanitized.size() == 0:
+		# State is completely invalid
+		_logger.warning("Received invalid state from %s for tick %s" % [sender_id, tick])
+		return
+
+	_states[tick] = PropertySnapshot.merge(_states.get(tick, {}), sanitized)
+	_latest_state_tick = tick
 		
-		if (NetworkRollback.enable_state_diffs and not _has_received_full_state):
-			_has_received_full_state = true
-			_receive_full_state_ack.rpc_id(1, received_tick)
-	else:
-		_logger.warning("Received invalid state from %s for tick %s" % [sender_id, received_tick])
+	if NetworkRollback.enable_state_diffs:
+		_receive_full_state_ack.rpc_id(get_multiplayer_authority(), tick)
 
-## The received state ALWAYS has missing properties.
 @rpc("any_peer", "unreliable_ordered", "call_remote")
-func _submit_diff_state(received_state: Dictionary, received_tick: int):
-	if received_tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state_tick >= 0:
+func _submit_diff_state(diff_state: Dictionary, tick: int, reference_tick: int):
+	if tick < NetworkTime.tick - NetworkRollback.history_limit:
 		# State too old!
-		_logger.error("Received state for %s, rejecting because older than %s frames" % [received_tick, NetworkRollback.history_limit])
+		_logger.error("Received diff state for %s, rejecting because older than %s frames" % [tick, NetworkRollback.history_limit])
 		return
 
-	if (_states.is_empty() || _latest_state_tick == -1):
-		_logger.error("Received a state without any properties from %s, but missing previous states!" % multiplayer.get_remote_sender_id())
-		return
+	if not _states.has(reference_tick):
+		# Reference tick missing, hope for the best
+		_logger.warn("Reference tick %d missing for %d" % [reference_tick, tick])
 
-	if (_states.has(received_tick - 1) == false):
-		_logger.info("non-linear diff states. Received tick %s while latest tick is %s" % [received_tick, _latest_state_tick])
-	
-	# Duplicates the previous state(s), including the current one
-	# Also fixes packet loss (e.g. a state missing)
-	for picked_tick in range(_latest_state_tick, received_tick):
-		_states[picked_tick + 1] = _states[picked_tick].duplicate(true)
```

Comment:
I will test this tomorrow (going to sleep now, its 00:58)
I remember that without this, packet losses lead to jittering and artifacts. It has been 2 weeks since I last updated this (how time flies) but I am certain that I had to duplicate past states to save packet losses (and even had some comments in the conversation tab, as to why I did this and also changed the `latest_state_tick` to -1

Reply:
> I remember that without this, packet losses lead to jittering and artifacts. 
I think that's because you assumed that state history was contiguous, so whenever a state was missing, you had to fill the gaps.

I've updated the PR to always diff from the last received full state, not the previous tick. So history gaps are not an issue.

Reply:
this change removes a big benefit of this PR, because this means a full state is sent for example every 0 to 128 ticks, but if I set it to 0 (that is the configuration I want for my game)

1. This PR will bug (I am testing this to confirm, give me an hour or so)
2. This PR doesn't allow for maximum bandwidth saving, as it forces full state sending every 128 ticks, instead of allowing the choice for the user to never send

--------------

For a complete review of the the additions, give me a few hours to confirm the above 2 and test for artifacts/jitters etc

Reply:
>I've updated the PR to always diff from the last received full state, not the previous tick. So history gaps are not an issue.

I want to offer a use-case example which is no longer possible if a missing diff property requires exclusively a full state property.
*Use-case*:
In every single tick the position property changes wildly. We have ticks 50 to 70 for this example. Time from host to client is 2 ticks. And 2 ticks from client to host.

The server sends full state position _[0,0]_ at tick 50.
It reaches the client at tick 52 (via rpc receive full state)
The client sends ack at tick 52, and it reaches the server at tick 54.

So the server also sends full state positions _[0,0]_ at tick 51, 52, 53, 54. Usual stuff so far.

The server sends diff state position _[9,0]_ at tick 55 (reaches the client at tick 57)
The server sends diff state position _[11,0]_ at tick 56 ( *the packet is lost* )
The server sends diff state position _[13,0]_ at tick 57 ( *the packet is lost* )
The server sends *empty* diff state position at tick 58 (reaches client at tick 60)

The client at tick 60, will then set the position it had received at tick 54 (full state tick) instead of what it received at tick 55.

----

Also, there is some bandwidth gain by being able to have exclusively diff states aka not sending full states between intervals

Reply:
I know, I understand why gaps in the state buffer would not work with always diffing with the previous state. That's why I've updated the PR to *not* do that. 

In the example above, even if you fill the gaps, you'd just end up copying the last known state, so I doubt the empty diff state would result in the expected state. 

I'm not convincet at all that diffing from the previous state would be reliable with anything but lossless network conditions. The best option I see is to ack diff states too. This way, only the initial full state send is required, the regular one can be disabled.

I'm open to implementing this in a separate PR. For now, I'd like to get the first version of diff states out so we can see how it performs.

Reply:
>In the example above, even if you fill the gaps, you'd just end up copying the last known state, so I doubt the empty diff state would result in the expected state.

The previous implementation would result in the expected state, because it bridged over the gaps gaps using the diffs. So for ticks 56, 57, 58, it would be a duplicate of state of tick 55 (which was a diff using full state of tick 54)

With this design choice implemented, I would argue against this PR because it makes the code more complicated, for negligible bandwidth gain, and also makes non-empty diffs followed up by empty diffs useless, since empty diffs will use full states (ignoring the non-empty diffs already received) :pensive: 

I think the final result with state diffs enabled (project settings) will have more jitters/artifacts than vanilla sending full states at all times, simply because some information is lost/ignored. This wasn't the case in the previous implementation, as both project settings had the same result

Reply:
This becomes more apparent, the more properties a state has.
For a quick use-case:
State of 4 properties (A,B,C,D)
At tick 50, full state tick is sent
At tick 51 diff state is sent with only A updated
At tick 52 diff state is sent with only B updated
At tick 53 diff state is sent with only C updated
At tick 54, 55, 56, empty states are sent, but packet loss happens
At tick 57 empty state is sent, but without packet loss

The received client on receiving the tick 57, should get a state with the following properties:
A of tick 51
B of tick 52
C of tick 53
D of tick 50

In current implementation
```gdscript
#reference_tick here is 50
var reference_state = _states.get(reference_tick, {})

if (diff_state.is_empty()):
		_latest_state_tick = tick
		_states[tick] = reference_state
```
the received client on receiving the tick 57, gets a state with the following properties:
A,B,C,D of tick 50

Reply:
I still do not feel you're familiar with this new implementation. Please make sure that you are before presenting any arguments.

> Use-case 2
> The received client on receiving the tick 57, should get a state with the following properties:

No. You've projected the data of the *original* implementation and ran it through the logic of the *new* implementation. That gets wrong results, but also doesn't lead anywhere. 

An empty state in the new implementation means that the actual state is the same as the reference state. 

--- 

Let's consider an example instead, with the following states:

```
tick 50: { a: 1, b: 2, c: 3 }
tick 51: { a: 8, b: 2, c: 3 }
tick 52: { a: 8, b: 9, c: 3 }
tick 53: { a: 8, b: 9, c: 7 }
tick 54: { a: 8, b: 9, c: 7 }
tick 55: { a: 8, b: 9, c: 7 }
tick 56: { a: 8, b: 9, c: 7 }
```

This would yield the following states sent:

```
tick 50, full state: { a: 1, b: 2, c: 3 }
tick 51, diff state from 50: { a: 8 }
tick 52, diff state from 50: { a: 8, b: 9 }
tick 53, diff state from 50: { a: 8, b: 9, c: 7 }
tick 54, diff state from 50: { a: 8, b: 9, c: 7 }
tick 55, diff state from 50: { a: 8, b: 9, c: 7 }
tick 56, diff state from 50: { a: 8, b: 9, c: 7 }
```

Remember, diffs are always calculated from the last full state sent. If let's say states 54 and 55 are lost due to packet loss, until state 56 arrives, the latest known state of 53 will be used.

---

Let's consider the same states, but now with always diffing with the previous state, as in the original implementation. The following states would be sent:


```
tick 50, full state: { a: 1, b: 2, c: 3 }
tick 51, diff state from 50: { a: 8 }
tick 52, diff state from 51: { b: 9 }
tick 53, diff state from 52: { c: 7 }
tick 54, diff state from 53: { }
tick 55, diff state from 54: { }
tick 56, diff state from 55: { }
```

If ticks 54 to 56 are lost, that's fine because nothing has changed. Instead, let's consider what happens when tick 52 is lost.

* Tick 50 arrives, state is now `{ a: 1, b: 2, c: 3 }`
* Tick 51 arrives, state is now `{ a: 8, b: 2, c: 3 }`
* Tick 52 is lost, state will be backfilled as `{ a: 8, b: 2, c: 3 }`
* Tick 53 arrives, state is now `{ a: 8, b: 2, c: 7 }`

The expected state for tick 53 is `{ a: 8, b: 9, c: 7 }`, however, due to a lost message, it is now `{ a: 8, b: 2, c: 7 }`. This is my issue with always diffing from the previous state, **as I have mentioned earlier.**

---

> I think the final result with state diffs enabled (project settings) will have more jitters/artifacts than vanilla sending full states at all times, simply because some information is lost/ignored. 

Strong disagree. States were already sent unreliably, gaps in the history buffer were always expected. If a state is unknown for a given tick, the latest known state is used. 

---

I'm inclined to merge this PR in its current state. One improvement we could do in the future is to ACK states more frequently over unreliable, so there's an option for doing smaller diffs.

Reply:
>One improvement we could do in the future is to ACK states more frequently over unreliable, so there's an option for doing smaller diffs.

I see this was included in a commit in this PR, nice.